src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -385,38 +377,40 @@ void GlobalizationNative_CloseSortHandle(SortHandle* pSortHandle)
         }
     }
 
-    pthread_mutex_destroy(&pSortHandle->collatorsLockObject);
-
     free(pSortHandle);
 }
 
 const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
 {
-    UCollator* pCollator;
     if (options == 0)
     {
-        pCollator = pSortHandle->collatorsPerOption[0];
+        return pSortHandle->collatorsPerOption[0];
     }
     else
     {
-        int lockResult = pthread_mutex_lock(&pSortHandle->collatorsLockObject);
-        if (lockResult != 0)
+        options &= CompareOptionsMask;
+        UCollator* pCollator = pSortHandle->collatorsPerOption[options];
+        if (pCollator != NULL)
         {
-            assert(FALSE && "Unexpected pthread_mutex_lock return value.");
+            return pCollator;
         }
 
-        options &= CompareOptionsMask;
-        pCollator = pSortHandle->collatorsPerOption[options];
-        if (pCollator == NULL)
+        pCollator = CloneCollatorWithOptions(pSortHandle->collatorsPerOption[0], options, pErr);
+
+#if HAVE_WORKING_STDATOMIC
+        UCollator* pNull = NULL;
+        if (!atomic_compare_exchange_strong((_Atomic(UCollator*)*)&pSortHandle->collatorsPerOption[options], &pNull, pCollator))
+#else
+        if (!__sync_bool_compare_and_swap(&pSortHandle->collatorsPerOption[options], NULL, pCollator))
```

Comment:
@jkotas if we are going to use __sync_bool_compare_and_swap in some cases, wouldn't be better just use it all the way and get rid of the ifdef's?

Reply:
If we going to use `__sync_bool_compare_and_swap` on ARM or other platforms, we may need to add extra barriers to make this work reliably.

Reply:
I would say that it's not architecture dependent but compiler version dependent.

I was able to build it using the `atomic_compare_exchange_strong` and Clang 6.0 on one of the JIT Team ARM machines. I think that we are hitting https://bugs.llvm.org/show_bug.cgi?id=37457

I can imagine that updating the Clang version we use is rather a big change and should be avoided?

@janvorli have you ever hit similar issue? I lack C expertise